src/main/scala/scalaz/ziokeeper/VectorClock.scala

```diff
@@ -0,0 +1,63 @@
+package scalaz.ziokeeper
+
+import scala.annotation.tailrec
+import VectorClock._
+case class VectorClock[A](value: Map[A, Long]) extends AnyVal {
+  def time(mid: A): Long = value.getOrElse(mid, 0L)
+
+  def increment(mid: A, value: Long): VectorClock[A] =
+    VectorClock(this.value.updated(mid, time(mid) + value))
+
+  def compare(right: VectorClock[A]): CompareVectorClock = {
+    val leftWithDefault  = this.value.withDefaultValue(0L)
+    val rightWithDefault = right.value.withDefaultValue(0L)
+
+    @tailrec
+    def loop(keys: List[A], leftBigger: Boolean, rightBigger: Boolean): (Boolean, Boolean) =
+      keys match {
+        case _ if leftBigger && rightBigger => (leftBigger, rightBigger)
+        case Nil                            => (leftBigger, rightBigger)
+        case h :: t =>
+          val diff = leftWithDefault(h) - rightWithDefault(h)
+          loop(t, leftBigger || diff > 0, rightBigger || diff < 0)
+      }
+
+    val keys = leftWithDefault.keySet.union(rightWithDefault.keySet)
+
+    loop(keys.toList, leftBigger = false, rightBigger = false) match {
+      case (false, false) => TheSame
+      case (true, true)   => Concurrent
+      case (true, false)  => After
+      case (false, true)  => Before
+    }
+  }
+
+  def merge(right: VectorClock[A]): VectorClock[A] = {
```

Comment:
Consider extracting out `mergeWith(right)((x: Long, y: Long) => Long)`, then you can implement this as `mergeWith(right)(_ max _)`.

Reply:
make sense

Reply:
done